udtube/data/conllu.py

```diff
@@ -3,25 +3,110 @@
 This is roughly compatible with the third-party package `conllu`, though it
 only has features we care about."""
 
+from __future__ import annotations
+
 import collections
+import dataclasses
 import re
 
-from typing import Dict, Iterable, Iterator, Optional, TextIO, Tuple
+from typing import Dict, Iterable, Iterator, List, Optional, TextIO, Tuple
+
+
+class Error(Exception):
+
+    pass
 
 
-# From: https://universaldependencies.org/format.html.
-_fieldnames = [
-    "id",
-    "form",
-    "lemma",
-    "upos",
-    "xpos",
-    "feats",
-    "head",
-    "deprel",
-    "deps",
-    "misc",
-]
+class ID:
+    """Representation of a CoNLL-U sentence ID.
+
+    Most of the time this is just a single integer, but MWEs use two integers
+    to represent a span of tokens.
+
+    Args:
+        lower (int): the lower or sole index.
+        upper (int, optional): the upper index, for MWEs.
+    """
+
+    lower: int
+    upper: int
+
+    def __init__(self, lower: int, upper: Optional[int] = None):
+        self.lower = lower
+        self.upper = self.lower if upper is None else upper
+
+    @classmethod
+    def parse_from_string(cls, string: str) -> ID:
+        if match := re.fullmatch(r"(\d+)-(\d+)", string):
+            return cls(int(match.group(1)), int(match.group(2)))
+        elif match := re.fullmatch(r"\d+", string):
+            return cls(int(match.group()))
+        else:
+            raise Error(f"Unable to parse ID {str}")
+
+    def __str__(self) -> str:
+        if self.is_mwe:
+            return f"{self.lower}-{self.upper}"
+        return str(self.lower)
+
+    def __len__(self) -> int:
+        return 1 + self.upper - self.lower
+
+    def __eq__(self, other: ID) -> bool:
+        return self.lower == other.lower and self.upper == other.upper
+
+    @property
+    def is_mwe(self) -> bool:
+        return len(self) > 1
+
+    def get_slice(self) -> slice:
+        # We add one to the right end since upper bounds are open in Python.
+        return slice(self.lower, self.upper + 1)
+
+
+@dataclasses.dataclass(slots=True)
+class Token:
+    """Token object."""
+
+    id_: ID  # Avoids clash with built-in.
+    form: str
+    lemma: str
+    upos: str
+    xpos: str
+    feats: str
+    head: str  # This could be parsed as an Optional[int] but YAGNI.
+    deprel: str
+    deps: str
+    misc: str
+
+    @classmethod
+    def parse_from_string(cls, string: str) -> Token:
+        id_, form, lemma, upos, xpos, feats, head, deprel, deps, misc = (
+            string.split("\t")
+        )
+        return cls(
+            ID.parse_from_string(id_),
+            form,
+            lemma,
+            upos,
+            xpos,
+            feats,
+            head,
+            deprel,
+            deps,
+            misc,
+        )
```

Comment:
I know it's good to be explicit and all, but couldn't we just

```
id_, *args = (string.split("\t"))
return cls(ID.parse_from_string(id_), *args)
```

Reply:
I'd almost perfer it. This feel like code bloat, it should be clear what cls contains from 71:80

Reply:
At least a few of others in the middle of the list will also be non-string types and will need similar conversions so this won't work. I will make `head` an `Optional[int]` in the next PR. Thus I would like to hold off on this because I think it will become a lot less attractive that way...